### ISSUE TITLE
JAX-2 — Padded static-shape jit pipeline, benchmark, and minimal JAX ASE calculator

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1590]
**Depends on:** JAX-1 (#1589) · **Blocks:** —

**Context:** JAX-1 (#1589) delivers mace-jax as an eager inference engine: `load_model` on the neutral checkpoint format, the block list for the three supported foundation models (MACE-MP-0 small, MACE-MPA-0 medium — the `mace_mp()` default — and MACE-OFF23 medium), E/F/stress outputs, and fp64 parity vs the committed torch goldens. Jit was deliberately *not* a JAX-1 (#1589) acceptance criterion, because under `jax.jit` every distinct `(n_nodes, n_edges)` shape recompiles — so in an MD trajectory every neighbor-list rebuild would recompile XLA mid-run. This is a *correctness-of-cost* problem, and the fix is measured on CPU: three raw sizes (10, 14, 23 nodes) → **3 traces**; the same three padded to one (64, 64) bucket → **1 trace**, with padded energies equal to unpadded (the self-loop padding edges sit at length `2·r_max`, where the polynomial cutoff is exactly 0 — `envelope * (x < r_max)`, `mace/modules/radial.py:128-142`, so the mask is a hard zero rather than a small number). This ticket delivers the padded static-shape jit pipeline over the project's one padding contract (DATA-3 (#1571); mirrored here as `mace_jax.graph`), the recompile-count gate that joins the mace-jax conformance suite, an inference benchmark, and a minimal JAX ASE calculator. The behavioural precedent is the torch calculator, which already implements this padding today: `_EDGE_PAD_MULTIPLE = 64` and `_EDGE_PAD_HEADROOM = 1.25` (`mace/calculators/mace.py:66-67`), bucket rounding via `_round_up` (`:70`), first-call auto-estimation (`:558-573`), grow-and-recompile on overflow (`:597-608`), unpadding via `_slice_real_outputs` (`:450-493`), and the padding graph itself in `mace/data/padding_tools.py:27-121`.

If capacity runs out, this ticket is an early de-scope candidate (JAX-1 (#1589) eager parity stays regardless); invoking that cut is an explicit, recorded deviation, never silent.

**Interface & constraints:**

1. **The padding contract (shared, not invented here — and already implemented in torch).** A padded batch is a *valid* batch: padding is physically inert *inside* the schema. `build_fake_padding_graph` (`mace/data/padding_tools.py:27-121`) is the pinned reference, and it does exactly this: one trailing fake graph holding the padding atoms with element-0 one-hot (`node_attrs[:, 0] = 1.0`), positions zeroed, `pbc` all-false, fake cell `eye(3) * max(2*r_max, 1)`; padding edges are self-loops on the **last** fake atom with `unit_shifts = e_x` and `shifts = 2·r_max · e_x`, so their length is `2·r_max` and the polynomial cutoff is exactly 0.0; `total_charge` zeroed and `total_spin` set to **ones**, not zeros; every other per-node/per-edge tensor zero-extended generically so a new graph key does not need an edit here. `batch`/`ptr` cover the **padded** sizes (so `num_graphs := ptr.numel()-1` and `num_nodes := positions.shape[0]` are the static shapes the compiler sees). Real counts live only in a sidecar:

```python
@dataclass(frozen=True)
class PaddingInfo:                    # mace_core.graph — framework-free
    n_real_nodes: int
    n_real_edges: int
    n_real_graphs: int

@dataclass(frozen=True)
class PaddingPolicy:
    mode: Literal["none", "fixed", "auto"] = "none"
    nodes_budget: int = 0
    edges_budget: int = 0
    graphs_budget: int = 0
    edge_multiple: int = 64      # legacy _EDGE_PAD_MULTIPLE (mace/calculators/mace.py:66)
    headroom: float = 1.25       # legacy _EDGE_PAD_HEADROOM (mace/calculators/mace.py:67)
    on_overflow: Literal["grow", "error"] = "grow"   # grow = bump budgets + exactly one retrace

# mace_jax.graph — the same interface DATA-3 (#1571) ships as mace_torch.graph:
def pad_batch(data: dict, policy: PaddingPolicy, r_max: float) -> tuple[dict, PaddingInfo]: ...
def unpad_outputs(outputs, info: PaddingInfo): ...   # graph-level -> [:n_real_graphs], node-level -> [:n_real_nodes]
```

Padding never reorders — it only appends — so `unpad ∘ pad = id` on real slots. The model, kernels, and `jax.jit` stay entirely padding-unaware: **no mask tensors enter the forward**. mace-jax requires `mode != "none"` on the jitted path.

2. **The stable jitted energy function (binding design constraint — keep jax-md a cheap follow-on, not v1.0).** The padded pipeline must expose a stable jitted `energy(positions, cell, graph)` whose tensor shapes are **invariant across neighbor-list rebuilds within a pad bucket** — i.e. the recompile-count gate (exactly 1 trace across real sizes at a fixed bucket) is the same property a jax-md energy function needs. Do not foreclose the jax-md neighbor-list/PBC seam: the padded neighbor contract (capacity + grow-with-one-recompile, above) must map onto jax-md's capacity/overflow model, so a future jax-md adapter is a **thin wrapper over this jitted energy fn, not a rewrite**. jax-md itself is out of v1.0. `positions` and `cell` are explicit differentiable arguments; everything else rides in the static-shape `graph` arrays.

3. **Inference-only derivative rules.** Forces/stress are *first* derivatives: `jax.grad` of the jitted energy, under jit, with **no `jax.custom_vjp` anywhere in the package** (pure-jax ops differentiate for free; there is no double-backward obligation in mace-jax, by construction). `segment_reduce` maps to `jax.ops.segment_sum` with `num_segments` supplied statically from the descriptor/graph — never inferred from `edge_index.max()` (empty-neighbourhood safety + host-sync rule); no shape anywhere on the jitted path may be read from data values.

4. **The recompile-count gate.** The jit-stability gate — recompilation count **exactly 1** across real structure sizes at a fixed pad bucket — activates with this ticket and joins the mace-jax conformance run (`run_jax_model_conformance`), per supported model per dtype (fp64 required + fp32). It was explicitly kept out of JAX-1 (#1589) acceptance; it lands here and must not be back-ported.

5. **The minimal JAX ASE calculator.** A thin ASE adapter over the jitted pipeline with `PaddingPolicy(mode="auto")`: result keys `energy`, `forces`, and `stress` when periodic; single-head runtime only (multi-head artifacts are rejected with a pointer to `mace model select-head`; a headless artifact is single-head); loads **only** the neutral checkpoint format — never legacy pickle, never importing `mace/` or `mace_torch`; a `clebsch_gordan_basis: "full"` artifact is rejected loudly, never silently projected. Within a bucket, MD steps re-enter the same trace; overflow grows the budgets and costs exactly one retrace — the same observable behaviour the torch calculator already logs at `mace/calculators/mace.py:597-608`. "Minimal" is scope, not a stub: it is the v1.0 JAX user surface (per-step Python; keeping the whole MD loop inside XLA is exactly what the future jax-md adapter adds).

6. **Benchmark.** A committed, CPU-runnable benchmark of the jitted padded pipeline over the supported model list: steady-state per-call latency after warm-up, jitted-padded vs eager-unpadded, with trace counts reported alongside timings (a benchmark that silently retraces measures compilation, not inference).

**Task:**
1. `mace_jax.graph`: `pad_batch` / `unpad_outputs` over `mace_core.graph`'s `PaddingInfo`/`PaddingPolicy` — the interface identical to DATA-3 (#1571)'s `mace_torch.graph`, producing `jax.Array` batches; the fake-graph construction reproduces `build_fake_padding_graph`'s pinned choices (element-0 one-hot, self-loops on the last padding atom, `unit_shifts = e_x`, cell `eye(3) * max(2·r_max, 1)`, `total_spin` ones); inertness test pins the padding-edge polynomial cutoff to exactly 0.0 at `2·r_max`.
2. The jitted pipeline: the stable `energy(positions, cell, graph)` plus jitted E/F/stress wrappers via `jax.grad`; forward is padding-unaware (no masks, no data-dependent shapes, static `num_segments`).
3. Trace-count instrumentation and the recompile-count gate: exactly 1 trace across ≥3 real sizes at a fixed bucket, wired into `run_jax_model_conformance` per supported model per dtype.
4. Padded-vs-unpadded conformance golden: outputs on real slots identical (per-dtype tolerance table) with and without padding, per supported model.
5. The minimal JAX ASE calculator (`mode="auto"`, grow-with-one-retrace), agreeing with the JAX-1 (#1589) eager results on the same configs.
6. The benchmark script, committed and CPU-runnable.

**Out of scope:** a jax-md adapter (future work; this ticket only keeps its seam open); LAMMPS-from-jax (reopening conditions are recorded elsewhere: this ticket landed + named demand + the XLA-in-LAMMPS memory story validated); any training or double-backward machinery (mace-jax is structurally inference-only); changes to JAX-1 (#1589)'s eager acceptance or to the torch-side padding implementation (DATA-3 (#1571)); growing the supported-model list.

**Acceptance criteria:**
- [ ] Trace count is exactly 1 across ≥3 real structure sizes at a fixed pad bucket, for each of the three supported models, at fp64 and fp32; the gate runs inside `run_jax_model_conformance`.
- [ ] Padded outputs on real slots are identical to unpadded outputs within the per-dtype tolerance table, per supported model; `unpad ∘ pad = id` on real slots.
- [ ] The padding graph matches the pinned torch construction field for field (element-0 one-hot, last-atom self-loops, `unit_shifts = e_x`, `shifts = 2·r_max · e_x`, `pbc` false, cell `eye(3) * max(2·r_max, 1)`, `total_charge` zero, `total_spin` one).
- [ ] Overflowing a bucket with `on_overflow="grow"` bumps the budgets and costs exactly one retrace; `"error"` raises loudly.
- [ ] The jitted `energy(positions, cell, graph)` signature is stable and shape-invariant within a bucket across neighbor-list rebuilds (asserted by rebuilding the neighbor list at fixed budgets and re-entering the same trace).
- [ ] `grep`-clean: no `jax.custom_vjp` in the package; no `edge_index.max()`-style shape inference on the jitted path; no mask tensors in the forward.
- [ ] The ASE calculator returns `energy`, `forces`, `stress`-when-periodic; agrees with the JAX-1 (#1589) eager path on the same configs at the per-dtype tolerance; rejects multi-head and full-basis artifacts loudly while loading headless artifacts; imports neither `mace/` nor `mace_torch`.
- [ ] The benchmark runs on CPU, reports steady-state timings with trace counts, and is committed.

**Verify:**
```bash
python -m pytest packages/mace-jax/tests/test_pad_batch.py -v            # inertness, unpad∘pad=id, grow-one-retrace, field-for-field vs the pinned construction
python -m pytest packages/mace-jax/tests/test_jit_pipeline.py -v         # recompile-count gate, padded==unpadded
python -m pytest packages/mace-jax/tests/test_ase_calculator.py -v       # keys, eager agreement, loud rejections
python packages/mace-jax/benchmarks/bench_jit_inference.py --device cpu
```
**Review focus:** the two properties that are this ticket's whole point — (1) the recompile count really is 1 (instrument traces; a benchmark or test that retraces per step is the failure mode this exists to kill), and (2) the jitted `energy(positions, cell, graph)` seam is genuinely wrappable by a future jax-md energy function (shape stability within a bucket, capacity/overflow semantics matching jax-md's model) — plus the exact-zero inertness of padding edges under jit at both dtypes, and that the fake-graph field choices were copied from the pinned torch construction rather than re-derived (`total_spin` ones and the last-atom self-loop are the two a rewrite gets wrong).

## Phase 5 — Electrostatics extensions, deployment, legacy retirement, release (15 tickets)

> **RET-1 (#1597)..RET-6 (#1602) are deletion-only PRs**: zero new v1 logic, budgeted only for deletion + parity-test degradation + CI config. Each retires one subsystem after it has been default-`v1` through its deprecation window with a green parity job (PAR-1 (#1572)), touches only `capabilities.toml` (INF-5 (#1554)) plus the deletion, removes the matching `xfail(strict=True)` fitness function, degrades its parity test from "in-process vs live legacy" to "vs frozen golden", and drives `git ls-files mace/` strictly downward — from **83** tracked files at `3b43d63f` (3 top-level + `data/` 8 + `tools/` 31 + `cli/` 17 + `calculators/` 10 + `modules/` 14) to 0. Order: energy → data → conversion CLIs → dipole/polar/magnetic → LAMMPS jit → god-module + launcher legacy branch. The 1.0 tag is hard-gated on `git ls-files mace/ == 0` (REL-2 (#1604)).
